source/opt/type_manager.cpp

```diff
@@ -347,10 +346,10 @@ void TypeManager::CreateDecoration(uint32_t target,
 }
 
 void TypeManager::RegisterType(uint32_t id, const Type& type) {
-  auto& t = id_to_type_[id];
-  t.reset(type.Clone().release());
-  if (GetId(t.get()) == 0) {
-    type_to_id_[t.get()] = id;
+  auto pair = type_pool_.insert(type.Clone());
```

Comment:
I think this needs a comment.
The comparison and hash on the  type pool will avoid inserting new clones of types that already match something already there.  In case of an attempted repeated insertion the clone will be destroyed by the end of this function.  That is, repeated insertions of the same Type will at most keep one corresponding object in the type pool..

Reply:
will fix